controller/lib/src/msvc/net_interface_imp.cpp

```diff
@@ -62,24 +62,62 @@ net_interface_imp::net_interface_imp()
             {
                 if (dev_addr->addr->sa_family == AF_INET && dev_addr->addr)
                 {
-                    char ip_str[INET_ADDRSTRLEN] = {0};
-                    struct sockaddr_in * sa = (struct sockaddr_in *)dev_addr->addr;
-                    inet_ntop(AF_INET, &sa->sin_addr, ip_str, INET_ADDRSTRLEN);
+                    char ip_str[INET_ADDRSTRLEN] = { 0 };
+                    inet_ntop(AF_INET, &((struct sockaddr_in *)dev_addr->addr)->sin_addr, ip_str, INET_ADDRSTRLEN);
                     device_ip_addresses.push_back(std::string(ip_str));
-                    
-                    uint64_t dev_mac;
-                    ULONG len;
-                    uint8_t tmp[16];
-                    len = sizeof(tmp);
-                    SendARP(sa->sin_addr.S_un.S_addr, INADDR_ANY, tmp, &len);
-                    utility::convert_eui48_to_uint64(&tmp[0], dev_mac);
-                    all_mac_addresses.push_back(dev_mac);
                 }
             }
-            
+
             all_ip_addresses.push_back(device_ip_addresses);
             total_devs++;
         }
+
+        /****************************** Find Adapter Info ***************************/
+        IP_ADAPTER_INFO * AdapterInfo;
+        ULONG AIS;
+        DWORD status;
+        AdapterInfo = (IP_ADAPTER_INFO *)calloc(total_devs, sizeof(IP_ADAPTER_INFO));
+        AIS = sizeof(IP_ADAPTER_INFO) * total_devs;
+
+        if (GetAdaptersInfo(AdapterInfo, &AIS) == ERROR_BUFFER_OVERFLOW)
+        {
+            free(AdapterInfo);
+            AdapterInfo = (IP_ADAPTER_INFO *)calloc(1, AIS);
+
+            if (AdapterInfo == NULL)
+            {
+                log_imp_ref->post_log_msg(LOGGING_LEVEL_ERROR, "Allocating memory needed to call GetAdaptersinfo.", dev->name);
+                return;
+            }
+
+            status = GetAdaptersInfo(AdapterInfo, &AIS);
+            if (status != ERROR_SUCCESS)
+            {
+                log_imp_ref->post_log_msg(LOGGING_LEVEL_ERROR, "GetAdaptersInfo call in netif_win32_pcap.c failed.", dev->name);
+                free(AdapterInfo);
+                return;
+            }
+        }
+
+        PIP_ADAPTER_INFO pAdapterInfo = NULL;
+        for (dev = all_devs; dev; dev = dev->next)
+        {
+            uint64_t dev_mac_addr = 0;
+            for (pAdapterInfo = AdapterInfo; pAdapterInfo != NULL; pAdapterInfo = pAdapterInfo->Next)
+            {
+                if (strstr(dev->name, pAdapterInfo->AdapterName) != 0)
+                {
+                    utility::MacAddr mac(
+                                         pAdapterInfo->Address[0], pAdapterInfo->Address[1],
+                                         pAdapterInfo->Address[2], pAdapterInfo->Address[3],
+                                         pAdapterInfo->Address[4], pAdapterInfo->Address[5]);
+                    dev_mac_addr = mac.tovalue();
```

Comment:
As an aside, there isn't a MacAddr constructor that takes an array of chars as an input?